src/arraymancer/linear_algebra/special_matrices.nim

```diff
@@ -47,3 +47,137 @@ proc vandermonde*[T](x: Tensor[T], order: int): Tensor[float] =
+  assert a.rank == 2
+  assert d.rank == 1
+  assert k < a.shape[0], &"Diagonal index ({k=}) exceeds matrix height ({a.shape[0]})"
+  assert k < a.shape[1], &"Diagonal index ({k=}) exceeds matrix width ({a.shape[1]})"
+  if anti:
+    if k >= 0:
+      when compileOption("boundChecks"):
+        let size = min(a.shape[0] - abs(k), a.shape[1])
+        assert size == d.size, &"Input size ({d.size}) does not match the {k}-th upper anti-diagonal size ({size})"
+      for i in 0 ..< d.size:
+        a[a.shape[0]-1-(i+k), i] = d[i]
+    else:
+      when compileOption("boundChecks"):
+        let size = min(a.shape[0], a.shape[1] - abs(k))
+        assert size == d.size, &"Input size ({d.size}) does not match the {-k}-th lower anti-diagonal size ({size})"
+      for i in 0 ..< d.size:
+        a[a.shape[0]-1-i, i-k] = d[i]
+  else:
+    if k >= 0:
+      when compileOption("boundChecks"):
+        let size = min(a.shape[0], a.shape[1] - abs(k))
+        assert size == d.size, &"Input size ({d.size}) does not match the {k}-th upper diagonal size ({size})"
+      for i in 0 ..< d.size:
+        a[i, i+k] = d[i]
+    else:
+      when compileOption("boundChecks"):
+        let size = min(a.shape[0] - abs(k), a.shape[1])
+        assert size == d.size, &"Input size ({d.size}) does not match the {-k}-th lower diagonal size ({size})"
+      for i in 0 ..< d.size:
+        a[i-k, i] = d[i]
+
+proc with_diagonal*[T](a: Tensor[T], d: Tensor[T], k = 0, anti = false): Tensor[T] {.noInit.}=
+  ## Copy the input matrix, changing one of its diagonals into the elements of the rank-1 input tensor d
+  ##
+  ## Input:
+  ##      - The matrix that will copied into the output.
+  ##      - Rank-1 tensor containg the elements that will be copied into the selected diagonal.
+  ##      - k: The index k of the diagonal that will be changed. The default is 0 (i.e. the main diagonal).
+  ##        Use k>0 for diagonals above the main diagonal, and k<0 for diagonals below the main diagonal.
+  ##      - anti: If true, set the k-th "anti-diagonal" instead of the k-th regular diagonal.
+  result = a
+  set_diagonal(result, d, k, anti=anti)
+
+proc identity*[T](n: int): Tensor[T] {.noInit.}=
+  ## Return an identity matrix (i.e. 2-D tensor) of size `n`
+  ## 
+  ## The identity matrix is a square 2-D tensor with ones on the main diagonal and zeros elsewhere.
+  ##
+  ## Input:
+  ##      - Number of rows / columns in the output.
+  ## Result:
+  ##      - The constructed indentity 2-D tensor
+  result = diag(ones[T](n))
+
+proc eye*[T](shape: varargs[int]): Tensor[T] {.noInit.}=
+  ## Return a 2-D tensor with ones on the diagonal and zeros elsewhere
+  ##
+  ## Input:
+  ##      - Rank-1 tensor containg the elements of the diagonal
+  ##      - The index of the diagonal that will be set. The default is 0.
+  ##        Use k>0 for diagonals above the main diagonal, and k<0 for diagonals below the main diagonal.
```

Comment:
The doc string seems wrong. And do I understand correctly that the difference between `eye` and `identity` is just that it allows to construct rectangular matrices? What happens with `shape.len > 2`?